File: 2019-03-04-496,503,556/lb/503.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        Stack<Integer> stack =  new Stack<>();
        int[] ret = new int[nums.length];
        for (int i=0;i<nums.length;i++){
            if (stack.size()==0 || nums[stack.peek()]>= nums[i]){
                stack.push(i);
            }else {
                while(stack.size()>0 && nums[stack.peek()] < nums[i]){
                    ret[stack.pop()] = nums[i];
                }
                stack.push(i);
            }
        }
        int i =0;
        while(i<nums.length){
            if (stack.size()==0)
                break;
            if (nums[i]>nums[stack.peek()])
                ret[stack.pop()] = nums[i];
            else
                i++;
        }
        while (stack.size() > 0)
            ret[stack.pop()] = -1;
        return ret;
    }
}

File: Year-2019/Mar/2019-03-04-496,503,556/watercow/503.py
# 骚操作：直接把nums扩充两倍，不用考虑循环
class Solution(object):
    def nextGreaterElements(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        res = [-1]*len(nums)
        stack = []
        for i in list(range(len(nums))) * 2:
            while stack and (nums[stack[-1]] < nums[i]):
                res[stack.pop()] = nums[i]
            stack.append(i)
        return res